include/adaflo/sharp_interface_util.h

```diff
@@ -0,0 +1,1108 @@
+
+        for (const auto q : fe_eval_dim.quadrature_point_indices())
+          {
+            Tensor<1, spacedim, double> result;
+            for (unsigned int i = 0; i < spacedim; ++i)
+              result[i] = -curvature_values[q] * normal_values[q][i] * fe_eval.JxW(q) *
+                          surface_tension;
+
+            integration_points.push_back(fe_eval.quadrature_point(q));
+            integration_values.push_back(result);
+          }
+      }
+  }
+
+  const auto [cells, ptrs, weights, points] = collect_integration_points(
+    dof_handler.get_triangulation(), mapping, integration_points, integration_values);
+
+  AffineConstraints<double> constraints; // TODO: use the right ones
+
+  FEPointEvaluation<spacedim, spacedim> phi_normal_force(mapping, dof_handler.get_fe());
+
+  std::vector<double>                  buffer;
+  std::vector<types::global_dof_index> local_dof_indices;
+
+  for (unsigned int i = 0; i < cells.size(); ++i)
+    {
+      typename DoFHandler<spacedim>::active_cell_iterator cell = {
+        &dof_handler.get_triangulation(), cells[i].first, cells[i].second, &dof_handler};
+
+      const unsigned int n_dofs_per_cell = cell->get_fe().n_dofs_per_cell();
+
+      local_dof_indices.resize(n_dofs_per_cell);
+      buffer.resize(n_dofs_per_cell);
+
+      cell->get_dof_indices(local_dof_indices);
+
+      const unsigned int n_points = ptrs[i + 1] - ptrs[i];
+
+      const ArrayView<const Point<spacedim>> unit_points(points.data() + ptrs[i],
+                                                         n_points);
+      const ArrayView<const Tensor<1, spacedim, double>> JxW(weights.data() + ptrs[i],
+                                                             n_points);
+
+      for (unsigned int q = 0; q < n_points; ++q)
+        phi_normal_force.submit_value(JxW[q], q);
+
+      phi_normal_force.integrate(cell, unit_points, buffer, EvaluationFlags::values);
+
+      constraints.distribute_local_to_global(buffer, local_dof_indices, force_vector);
+    }
+}
+
+
+
+template <int dim, int spacedim, typename VectorType>
+void
+compute_normal(const Mapping<dim, spacedim> &   mapping,
+               const DoFHandler<dim, spacedim> &dof_handler_dim,
+               VectorType &                     normal_vector)
+{
+  FEValues<dim, spacedim> fe_eval_dim(mapping,
+                                      dof_handler_dim.get_fe(),
+                                      dof_handler_dim.get_fe().get_unit_support_points(),
+                                      update_normal_vectors | update_gradients);
+
+  Vector<double> normal_temp;
+
+  for (const auto &cell : dof_handler_dim.get_triangulation().active_cell_iterators())
+    {
+      TriaIterator<DoFCellAccessor<dim, spacedim, false>> dof_cell_dim(
+        &dof_handler_dim.get_triangulation(),
+        cell->level(),
+        cell->index(),
+        &dof_handler_dim);
+
+      fe_eval_dim.reinit(dof_cell_dim);
```

Comment:
Do you need this? It seems you could simply do
```suggestion
      fe_eval_dim.reinit(cell);
```